File: 08.js
import { readFile } from './tools.ts';

/** @typedef {["acc"|"nop"|"jmp", number]} Instruction */

/** @type {Instruction[]} */
const input = readFile('inputs/08.txt').split('\r\n').map(line => {
  const stuff = line.split(' ');
  return [stuff[0], parseInt(stuff[1])];
});

/**
 * @param {Instruction[]} instructions
 * @returns {[boolean, number]}
 */
const executor = (instructions) => {
  let position = 0;
  let acc = 0;
  const seen = new Set();
  const functions = {
    nop: () => position++,
    jmp: arg => position += arg,
    acc: arg => {
      acc += arg;
      position++;
    },
  }
  while(position < instructions.length){
    if(seen.has(position) || position < 0) return [false, acc];
    seen.add(position);
    const [instruction, arg] = instructions[position];
    functions[instruction](arg);
  }
  return [true, acc];
}

const toggle = str => str === 'jmp' ? 'nop' : 'jmp';

console.log(executor(input)[1]);

for(let i = 0; i < input.length; i++){
  if(input[i][0] === 'acc') continue;
  input[i][0] = toggle(input[i][0]);
  const [graceful, acc] = executor(input);
  if(graceful) console.log(acc);
  input[i][0] = toggle(input[i][0]);
}

File: otherLangs/haskell/08.hs
import Control.Monad ()
import Data.List.Split ( splitOn )
import qualified Data.Set as Set

main :: IO ()
main = do
  contents <- readFile "inputs/08.txt"
  let input = map parseLine $ lines contents
  print $ executor input 0 0 Set.empty

executor :: [([Char], Int)] -> Int -> Int -> Set.Set Int -> Int
executor tape acc position seen = if position `elem` seen then acc else case tape !! position of
  ("acc", arg) -> executor tape (acc + arg) (position + 1) (Set.union seen $ Set.singleton position)
  ("jmp", arg) -> executor tape acc (position + arg) (Set.union seen $ Set.singleton position)
  ("nop", _) -> executor tape acc (position + 1) (Set.union seen $ Set.singleton position)


parseLine ::[Char] -> ([Char], Int)
parseLine line = (head split, read $ filter (/='+') $ split !! 1)
  where
    split = splitOn " " line
